src/yaml/from_yaml.cr

```diff
@@ -22,39 +29,110 @@ end
+      value.to_i{{size.id}}(prefix: true, underscore: true, strict: true)
+    rescue ex
+      raise YAML::ParseException.new(ex.message.not_nil!, *location)
+    end
+  end
+
+  def UInt{{size.id}}.new(pull : YAML::PullParser)
+    location = pull.location
+    value = pull.read_plain_scalar
+    begin
+      value.to_u{{size.id}}(prefix: true, underscore: true, strict: true)
     rescue ex
       raise YAML::ParseException.new(ex.message.not_nil!, *location)
     end
   end
 {% end %}
 
-def String.new(pull : YAML::PullParser)
-  pull.read_scalar
-end
+{% for size in %w(32 64) %}
+  def Float{{size.id}}.new(pull : YAML::PullParser)
+    location = pull.location
+    begin
+      case value = pull.read_plain_scalar
+      when ".inf", ".Inf", ".INF", "+.inf", "+.Inf", "+.INF"
+        INFINITY
+      when "-.inf", "-.Inf", "-.INF"
+        -INFINITY
+      when ".nan", ".NaN", ".NAN"
+        NAN
+      else
+        value.sub('_', "").to_f{{size.id}}
+      end
+    rescue ex
+      raise YAML::ParseException.new(ex.message.not_nil!, *location)
+    end
+  end
+{% end %}
 
-def Float32.new(pull : YAML::PullParser)
-  pull.read_scalar.to_f32
+def Time.new(pull : YAML::PullParser)
+  location = pull.location
+  value = pull.read_plain_scalar
+
+  # Coerce into a format we can parse with Time.parse
+  date_time_parts = value.squeeze(" ", "\t").upcase.tr("\t", " ").tr(" ", "T").split('T', 2)
+  date = date_time_parts[0]?
+  time_with_zone = date_time_parts[1]?.to_s.sub('-', " -").sub('+', " +").sub('Z', " Z")
+  time_with_zone_parts = time_with_zone.split(' ', 2)
+  time = time_with_zone_parts[0]?.to_s.split('.', 2)[0]?.to_s.rchop('T')
+  zone_parts = time_with_zone_parts[1]?.to_s.split(':', 2)
+  zone = if zone_parts[0]? == "Z"
+           zone_parts[0]
+         elsif !zone_parts[0]?.to_s.empty?
+           zone_sign = zone_parts[0]?.to_s[0]
+           zone_hour = zone_parts[0]?.to_s[1..-1].rjust(2, '0')
+           zone_min = zone_parts[1]? || "00" unless zone_hour.size == 5
+           zone_sign + [zone_hour, zone_min].compact.join(':')
+         end
+  time_with_zone = [time, zone].join
+  value = [date, time_with_zone].compact.join('T')
+
+  # Loop over the supported formats until we can parse one
+  formats = %w(%Y %Y-%m %F) + %w(%FT%H %FT%H:%M %FT%H:%M:%S %FT%H:%M:%S% %FT%H:%M:%S% %FT%X%z)
+  until formats.empty?
+    current_format = formats.pop
+    begin
+      return Time::Format.new(current_format).parse(value)
+    rescue Time::Format::Error
+    end
+  end
+  raise YAML::ParseException.new("Could not parse time from '#{value}'", *location)
 end
 
-def Float64.new(pull : YAML::PullParser)
-  pull.read_scalar.to_f64
+def String.new(pull : YAML::PullParser)
+  location = pull.location
+  return pull.read_scalar unless pull.data.scalar.style == LibYAML::ScalarStyle::PLAIN
+  pull.read_plain_scalar.tap do |value|
+    if reserved = [Nil, Bool, Int64, Float64, Time].find(&.from_yaml? "---\n#{value}")
```

Comment:
okay... so duplication, no duplication... i'm getting pulled in two directions here. 😞